### PR TITLE
Move review stack resource allocation queue into shipit-engine

### PR DESCRIPTION
**Pull request by indiebrain**
In order to include it in the upstream pull request, we want to move
the "ResourceAllocationQueue" from Milano into the shipit-engine
project.

Refrerences
-----------

- https://nitro.powerhrg.com/runway/backlog_items/GOG-216

### CODE REVIEW

File: test/models/shipit/review_stack_test.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    end
+
+    test ".review_request returns latest non merge request" do
+      stack = shipit_stacks(:review_stack)
+      pull_request = PullRequest.create!(stack: stack, number: "1", review_request: true)
+
+      assert stack.review_request, pull_request
+    end
+
+    test "automatically enqueues a for provisioning only on create" do
```
Comment (benlangfeld) [inline test/models/shipit/review_stack_test.rb:25]:
```suggestion
    test "automatically enqueues for provisioning only on create" do
```

Yes I read the tests in code review 😝

Comment (indiebrain) [reply test/models/shipit/review_stack_test.rb]: Just checking 😀

**Comment by indiebrain** [comment]
🚨  Pump the breaks! I don't thinks this should be merged as is. I believe adding this bit

https://github.com/powerhome/shipit-engine/blob/b4e37e81a9f2c54afc346a9c716923c2c0c2a0ed/app/models/shipit/review_stack.rb#L6

to the `ReviewStack` is preventing the `github_sync` `after_commit` callback from the parent `Stack` class.

https://github.com/powerhome/shipit-engine/blob/8c36dbd96edf0cc39e100efd586f11031209450a/app/models/shipit/stack.rb#L81

 In the host application the ReviewStack sits re-trying to sync its PullRequest, but failing because the ReviewStack hasn't synced any commits yet.

**Comment by indiebrain** [comment]
@benlangfeld this feels like a deeper issue with [ActiveRecord Callbacks](https://github.com/rails/rails/blob/v6.0.3/activesupport/lib/active_support/callbacks.rb) and chaining. I've spent the past 2 hours trying to chase it down, but find myself no closer to an explanation. I propose we roll this body of work back to https://github.com/powerhome/shipit-engine/pull/43/commits/2538645b7a0b15b6fb228324305128373b2ae80a and deal with automatically enqueueing the `ReviewStack` via lifecycle hooks later.

What do you think?

**Pull request merged**